### PR TITLE
Feature/cron events

### PR DESCRIPTION
**Pull request by androsanta**
Wip - Initial implementation for scheduled events

todo:
- [x] Misc.createAwsSafeFunctionName should strip symbols (-./ etc...)
- [x] generate exporter file (as well as source files), should be a different file from the endpoints exporter right?
- [x] tests

### CODE REVIEW

File: packages/restlessness-core/src/Misc/index.ts [hunk 1]
```diff
@@ -36,4 +37,25 @@ export default class Misc {
       }
     }
   };
+  static createAwsSafeFunctionName = (functionId: string, fullServiceName: string) => {
+    const id = functionId; //@TODO check for symbols presence (-./ etc..)
```
Comment (androsanta) [inline packages/restlessness-core/src/Misc/index.ts:41]: is it ok to implement this here?

File: packages/restlessness-core/assets/boilerplate/configs/schedule-events.json [hunk 2]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (antoniogiordano) [inline packages/restlessness-core/assets/boilerplate/configs/schedule-events.json:1]: Let's choose a single name like endpoints, event is generic, maybe schedules.json or crons.json can work

File: packages/restlessness-core/src/JsonScheduleEvents/index.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import JsonConfigFile, { JsonConfigEntry } from '../JsonConfigFile';
```
Comment (antoniogiordano) [inline packages/restlessness-core/src/JsonScheduleEvents/index.ts:1]: Idem here, JsonSchedules or JsonCrons

File: packages/restlessness-core/src/JsonScheduleEvents/index.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  description?: string
+  rate: string
+  enabled?: boolean
+  serviceName: string
+  input?: {
+    type: 'input' | 'inputPath' | 'inputTransformer'
+    value: { [key: string]: any }
+  }
+  safeFunctionName: string
+}
```
Comment (antoniogiordano) [inline packages/restlessness-core/src/JsonScheduleEvents/index.ts:22]: I think it should have daos array as well

File: packages/restlessness-core/src/Misc/index.ts (hunk 1 above)
Comment (antoniogiordano) [reply packages/restlessness-core/src/Misc/index.ts:41]: It should be fine!

**Review by antoniogiordano — CHANGES_REQUESTED** [review]

File: packages/restlessness-core/src/JsonScheduleEvents/index.ts (hunk 4 above)
Comment (androsanta) [reply packages/restlessness-core/src/JsonScheduleEvents/index.ts:22]: We should modify also the LambdaHandler (currently it handles only endpoints) or define another handler expecially for schedule events. This way daos packages hooks will be executed

Comment (antoniogiordano) [reply packages/restlessness-core/src/JsonScheduleEvents/index.ts]: Correct, I was thinking something about a ScheduleHandler or CronHandler :-)

Comment (antoniogiordano) [reply packages/restlessness-core/src/JsonScheduleEvents/index.ts]: Ok ScheduleHandler 👍

**Comment by androsanta** [comment]
Also imho it could be useful to split the JsonServices class into sub-classes to simplify it, what do you think @antoniogiordano ?

I have something in mind, I will draft it after the ScheduleHandler task

**Comment by antoniogiordano** [comment]
@androsanta ok sure, I realised also is a little bit confusing right now

**Comment by androsanta** [comment]
Serverless-offline introduced support for schedule events on version 6.x while we currently use version 5.x
I'll check if there are major changes that affect the way we use it

File: packages/restlessness-cli/src/commands/dev/index.ts [hunk 5]
```diff
@@ -42,7 +42,7 @@ The ChildProcess object is returned through a promise, with some callbacks alrea
 
 function spawnBackend(): Promise<ChildProcess> {
   return new Promise((resolve, reject) => {
-    const proc = spawn('serverless', ['offline', '--config', 'serverless-services/offline.json', '--port', '4123'], {
+    const proc = spawn('serverless', ['offline', '--config', 'serverless-services/offline.json', '--httpPort', '4123', '--lambdaPort', '3003'], {
```
[the remark this answers is not in the document]
Comment (androsanta) [reply packages/restlessness-cli/src/commands/dev/index.ts:45]: It's explained [here](https://github.com/dherault/serverless-offline/issues/987#issuecomment-680871859)

File: packages/restlessness-core/src/JsonSchedules/templates/index.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+export {
+  ${jsonScheduleEventsEntries.map(jsonScheduleEventsEntry => `${jsonScheduleEventsEntry.safeFunctionName},`).join('\n  ')}
+};
+`;
+
+export const indexTemplate = (scheduleId: string) => `import { ScheduleHandler } from '@restlessness/core';
+import handler from './handler';
+
+export default ScheduleHandler.bind(this, handler, '${scheduleId}');
```
[the remark this answers is not in the document]
Comment (androsanta) [reply packages/restlessness-core/src/JsonSchedules/templates/index.ts:14]: Yes, it works the same way as the LambdaHandler. The ScheduleHandler function needs the scheduleId to get the JsonScheduleEntry from the config file

File: packages/restlessness-core/src/Misc/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      expect(isSafe(safeFunctionName)).toBe(true);
+    });
+
+    expect(
+      () => Misc.createAwsSafeFunctionName('10/()$23', 'service-test-1'),
+    ).toThrowError();
+
+    done();
+  });
+});
```
[the remark this answers is not in the document]
Comment (antoniogiordano) [reply packages/restlessness-core/src/Misc/index.test.ts:25]: Exactly, we should check that the cut is right and final length is <= 64

File: packages/restlessness-core/src/JsonSchedules/templates/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (antoniogiordano) [reply packages/restlessness-core/src/JsonSchedules/templates/index.ts:14]: Ok but this '${scheduleId}' should be with the others apex `${scheduleId}`?

[the remark this answers is not in the document]
Comment (androsanta) [reply packages/restlessness-core/src/JsonSchedules/templates/index.ts:14]: The id is a string, as in the [JsonEndpoint indexTemplate](https://github.com/getapper/restlessness/blob/680fbf63bb03d1c435d824e00c991573d2bd4679/packages/restlessness-core/src/JsonEndpoints/templates/index.ts#L11)

**Review by antoniogiordano — APPROVED** [review]

**Pull request merged**